packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo

```diff
@@ -0,0 +1,208 @@
+        +ERC721Component::ERC721HooksTrait<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of IERC721Enumerable<ComponentState<TContractState>> {
+        /// Returns the total amount of tokens stored by the contract.
+        fn total_supply(self: @ComponentState<TContractState>) -> u256 {
+            self.ERC721Enumerable_all_tokens_len.read()
+        }
+
+        /// Returns a token id at a given `index` of all the tokens stored by the contract.
+        /// Use along with `total_supply` to enumerate all tokens.
+        ///
+        /// Requirements:
+        ///
+        /// - `index` is less than the total token supply.
+        fn token_by_index(self: @ComponentState<TContractState>, index: u256) -> u256 {
+            assert(index < self.total_supply(), Errors::OUT_OF_BOUNDS_INDEX);
+            self.ERC721Enumerable_all_tokens.read(index)
+        }
+
+        /// Returns the token id owned by `owner` at a given `index` of its token list.
+        /// Use along with `ERC721::balance_of` to enumerate all of `owner`'s tokens.
+        ///
+        /// Requirements:
+        ///
+        /// - `index` is less than `owner`'s token balance.
+        /// - `owner` is not the zero address.
+        fn token_of_owner_by_index(
+            self: @ComponentState<TContractState>, owner: ContractAddress, index: u256
+        ) -> u256 {
+            let erc721_component = get_dep_component!(self, ERC721);
+            assert(index < erc721_component.balance_of(owner), Errors::OUT_OF_BOUNDS_INDEX);
+            self.ERC721Enumerable_owned_tokens.read((owner, index))
+        }
+    }
+
+    //
+    // Internal
+    //
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl ERC721: ERC721Component::HasComponent<TContractState>,
+        +ERC721Component::ERC721HooksTrait<TContractState>,
+        impl SRC5: SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Initializes the contract by declaring support for the `IERC721Enumerable`
+        /// interface id.
+        fn initializer(ref self: ComponentState<TContractState>) {
+            let mut src5_component = get_dep_component_mut!(ref self, SRC5);
+            src5_component.register_interface(interface::IERC721ENUMERABLE_ID);
+        }
+
+        /// Updates the ownership and token-tracking data structures.
+        ///
+        /// When a token is minted (or burned), `token_id` is added to (or removed from)
+        /// the token-tracking structures.
+        ///
+        /// When a token is transferred, minted, or burned, the ownership-tracking data structures
+        /// reflect the change in ownership of `token_id`.
+        ///
+        /// This must be added to the implementing contract's `ERC721HooksTrait::before_update`
+        /// hook.
+        fn before_update(
```

Comment:
In the Solidity version, this is called after the update, not before. Let us refactor this to match it at least for consistency.

Reply:
You are correct; however, this will break the design of the current hooks implementation. In the Solidity version, the extension overrides `_update` and first calls `super._update` which returns the previous owner. We can't do this because we're hooking into `update` and there's no `from` param (we're getting `from` from `_owner_of`). As an example, minting a token will not update the total supply because the previous owner will be the token recipient and not the zero address

We can consider a bigger refactor with erc721 if it makes sense to add `from` to `update`

Note also that prior to v5.0.0 in the solidity lib, this extension used `_beforeTokenTransfer`

Reply:
Got it. As you said is better to leave it this way. No need to refactor because of this.